### ISSUE TITLE
New analyzer: Do not assign new value to a parameter if not ref/out

### ISSUE THREAD

**Issue by Evangelink** [dotnet/roslyn-analyzers#3568]
The idea behind this analyzer is both for readability and also to avoid "wrong" behavior that some junior devs may assume coming from other languages. The rule would recommend creating a local variable to make it clear this won't actually change the parameter reference outside the scope of the method.

```csharp
public class C
{
	public void M1(int i)
	{
		i = i + 2; // Diagnostic
	}

	public void M2(C c)
	{
		c = null; // Diagnostic
	}

	public void M3(ref int i, ref C c)
	{
		i = 42;   // ok
		c = null; // ok
	}

	public void M3(out int i, out C c)
	{
		i = 42;   // ok
		c = null; // ok
	}
}
```

**Comment by mavasani** [comment]
VS IDE will flag such parameters with IDE0060. Again available for build enforcement through https://dotnet.myget.org/feed/roslyn/package/nuget/Microsoft.CodeAnalysis.CSharp.CodeStyle.

**Comment by Evangelink** [comment]
@mavasani `IDE0060` only raise in the case of `M2`. Maybe I need to create a ticket on roslyn side.

**Comment by mavasani** [comment]
@Evangelink I am not sure if we want a diagnostic for M1. The input value is indeed read to compute a new value, and the new value would likely be used later in the program (otherwise IDE0059 would be generated for unused value assignment to i).

**Comment by Evangelink** [comment]
The idea of my analyzer wasn't to catch reading before assigning (that's another rule suggestion). I have seen some devs who thought that `i` would be changed outside the method.

**Comment by mavasani** [comment]
> The idea of my analyzer wasn't to catch reading before assigning (that's another rule suggestion). I have seen some devs who thought that i would be changed outside the method.

Unfortunately, I feel that would be a pretty noisy analyzer. How do you distinguish a case where the assignment was indeed desirable to _repurpose_ the parameter. For example:

```csharp
void M(string str)
{
   // Append a suffix to create a new string for further processing
   str = str + "SomeSuffix";

  ... code which reads 'str' and expects to end with "SomeSuffix" ...

}
```

We need input parameter as a building block, and don't want to define a new local variable for first assignment to ensure the subsequent code does not accidentally read the value passed from the caller. Assigning to `str` in first line essentially masks the input seed value.

**Comment by Evangelink** [comment]
> How do you distinguish a case where the assignment was indeed desirable to repurpose the parameter? 

You don't! The rule idea is really to say let's assume non `ref` / `out` parameters are readonly. So in your case you don't reassign but you create a new variable instead. That's more of a styling rule.

**Comment by Evangelink** [comment]
If I read correctly in between the line, I would say that you are not so much interested in this rule. Shall I close it?

**Comment by sharwell** [comment]
This seems like it would be a very noisy rule. I believe the current unused value analysis covers the cases of more importance; if a method is long enough that an assignment to the parameter is a problem, then it's likely the method length that's a problem and not the reassignment.

**Comment by HaloFour** [comment]
@sharwell 

> This seems like it would be a very noisy rule

I'm poking around on this repo as it seems that the proposal to add support for `readonly` parameters/locals to C# has hit a bit of a dead end and it has been suggested that such guardrails are better enforced by analyzers.  I'd like to propose such an analyzer, which is similar to this one but would cover many more use cases, but such an analyzer would be necessarily noisy even if largely "disabled"/"suppressed" by default.  So before I dig into a proposal I wanted to get a sense of what the members of this repo thought about the subject.

**Comment by RikkiGibson** [comment]
Hi @HaloFour, I didn't see your message at the time, but I would be willing to at least review and give feedback on such a proposal. I think some kind of analyzer solution is probably most realistic for getting/simulating a "readonly locals" feature.

**Comment by HaloFour** [comment]
@RikkiGibson Would it be preferable to carry out that conversation on this issue/discussion or to open another one?

**Comment by RikkiGibson** [comment]
I would open a new issue for the new proposal.

**Comment by lawrence-laz** [comment]
Was the new issue ever opened?

**Comment by RikkiGibson** [comment]
#7422
